File: ios/MendixNative/NativeFsModule/NativeFsModule.m
#import "NativeFsModule.h"
#import "ReactNative.h"
#import "SSZipArchive.h"
#import <Foundation/Foundation.h>
#import <React/RCTBlobManager.h>
#import <React/RCTEventEmitter.h>

BOOL _enableEncryption;

@implementation NativeFsModule

RCT_EXPORT_MODULE()

NSString *ERROR_SAVE_FAILED = @"ERROR_SAVE_FAILED";
NSString *ERROR_READ_FAILED = @"ERROR_READ_FAILED";
NSString *ERROR_MOVE_FAILED = @"ERROR_MOVE_FAILED";
NSString *ERROR_DELETE_FAILED = @"ERROR_DELETE_FAILED";
NSString *ERROR_SERIALIZATION_FAILED = @"ERROR_SERIALIZATION_FAILED";
NSString *INVALID_PATH = @"INVALID_PATH";

NSErrorDomain const NativeFsErrorDomain =
    @"com.mendix.mendixnative.nativefsmodule";

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

+ (void)setEncryptionEnabled:(BOOL)e {
    _enableEncryption = e;
}

+ (NSString *)formatError:(NSString *)message {
    return [[[NativeFsModule description] stringByAppendingString:@": "]
        stringByAppendingString:message];
}

+ (NSData *)readBlobRefAsData:(NSDictionary<NSString *, id> *)blob {
    RCTBlobManager *blobManager =
        [[ReactNative.instance getBridge] moduleForClass:RCTBlobManager.class];
    return [blobManager resolve:blob];
}

+ (NSDictionary *)readDataAsBlobRef:(NSData *)data {
    RCTBlobManager *blobManager =
        [[ReactNative.instance getBridge] moduleForClass:RCTBlobManager.class];

    return @{
        @"blobId" : [blobManager store:data],
        @"offset" : @0,
        @"length" : @(data.length)
    };
}

+ (NSData *)readData:(NSString *)filePath {
    if (![[NSFileManager defaultManager] fileExistsAtPath:filePath]) {
        return nil;
    }

    NSError *error;
    NSData *data =
        [NSData dataWithContentsOfURL:[NSURL fileURLWithPath:filePath]
                              options:NSDataReadingMappedAlways
                                error:&error];
    if (error) {
        return nil;
    }

    return data;
}

+ (NSDictionary *)readJson:(NSString *)filePath
                     error:(NSError *_Nullable *)error {
    NSData *data = [NativeFsModule readData:filePath];
    if (data == nil) {
        return nil;
    }
    
    NSDictionary *res = [NSJSONSerialization JSONObjectWithData:data
                                                        options:NSJSONReadingMutableContainers
                                                          error:error];
    if (error && *error) {
        return nil;
    }

    return res;
}

+ (BOOL)save:(NSData *)data
    filepath:(NSString *)filepath
       error:(NSError *_Nullable *)error {
    NSURL *directoryUrl =
        [NSURL fileURLWithPath:[filepath stringByDeletingLastPathComponent]];
    [[NSFileManager defaultManager] createDirectoryAtURL:directoryUrl
                             withIntermediateDirectories:YES
                                              attributes:nil
                                                   error:error];
    if (error && *error) {
        return false;
    }
    
    NSDataWritingOptions options = NSDataWritingAtomic;
    if (_enableEncryption) {
        options = NSDataWritingAtomic | NSDataWritingFileProtectionComplete;
    }

    [data writeToURL:[NSURL fileURLWithPath:filepath]
             options:options
               error:error];
    if (error && *error) {
        return false;
    }

    return true;
}

+ (BOOL)move:(NSString *)filepath
     newPath:(NSString *)newPath
       error:(NSError **)error {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:filepath]) {
        *error = [NSError
            errorWithDomain:NativeFsErrorDomain
                       code:-1
                   userInfo:@{
                       NSLocalizedDescriptionKey : @"File does not exist"
                   }];
        return false;
    }

    NSURL *directoryUrl =
        [NSURL fileURLWithPath:[newPath stringByDeletingLastPathComponent]];

    [[NSFileManager defaultManager] createDirectoryAtURL:directoryUrl
                             withIntermediateDirectories:YES
                                              attributes:nil
                                                   error:error];
    if (error && *error) {
        return false;
    }

    [fileManager moveItemAtPath:filepath toPath:newPath error:error];
    if (error && *error) {
        return false;
    }

    return true;
}

+ (BOOL)remove:(NSString *)filepath error:(NSError **)error {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:filepath]) {
        return false;
    }

    [fileManager removeItemAtPath:filepath error:error];
    if (error && *error) {
        return false;
    }
    return true;
}

+ (BOOL)ensureWhiteListedPath:(NSArray<NSString *> *)paths
                        error:(NSError **)error {
    for (id path in paths) {
        if (!([path hasPrefix:[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject]] ||
            [path hasPrefix:[NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject]] || [path hasPrefix:[NSTemporaryDirectory() stringByStandardizingPath]])) {
            *error = [NSError
                errorWithDomain:NativeFsErrorDomain
                           code:999
                       userInfo:@{
                           NSLocalizedDescriptionKey : [NSString
                               stringWithFormat:@"The path %@ does not point "
                                                @"to the documents directory",
                                                path]
                       }];
            return false;
        }
    }
    return true;
}

+ (NSArray<NSString *> *)list:(NSString *)dirPath {
    NSArray<NSString *> *listedFiles =
        [[[NSFileManager defaultManager] enumeratorAtPath:dirPath] allObjects];
    return listedFiles == nil ? @[] : listedFiles;
}

RCT_EXPORT_METHOD(setEncryptionEnabled
                  : (BOOL) enabled
                  ) {
    [NativeFsModule setEncryptionEnabled:enabled];
}

RCT_EXPORT_METHOD(save
                  : (NSDictionary<NSString *, id> *)blob filepath
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    NSData *data = [NativeFsModule readBlobRefAsData:blob];
    [NativeFsModule save:data filepath:filepath error:&error];
    if (error) {
        return reject(ERROR_SAVE_FAILED,
                      [NativeFsModule formatError:@"Save failed"], error);
    }

    resolve(nil);
}

RCT_EXPORT_METHOD(read
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSData *data;
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    data = [NativeFsModule readData:filepath];
    if (data == nil) {
        return resolve(nil);
    }
    
    NSDictionary<NSString *, id> *blob =
        [NativeFsModule readDataAsBlobRef:data];

    resolve(blob);
}

RCT_EXPORT_METHOD(move
                  : (NSString *)filepath
                  : (NSString *)newPath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObjects:filepath, newPath, nil]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    [NativeFsModule move:filepath newPath:newPath error:&error];
    if (error) {
        return reject(ERROR_MOVE_FAILED,
                      [NativeFsModule formatError:@"Failed to move file"],
                      error);
    }

    resolve(nil);
}

RCT_EXPORT_METHOD(remove
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    [NativeFsModule remove:filepath error:&error];
    if (error) {
        return reject(ERROR_DELETE_FAILED,
                      [NativeFsModule formatError:@"Failed to delete file"],
                      error);
    }

    resolve(nil);
}

RCT_EXPORT_METHOD(list
                  : (NSString *)dirPath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:dirPath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }
    resolve([NativeFsModule list:dirPath]);
}

RCT_EXPORT_METHOD(readAsDataURL
                  : (NSString *)filePath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filePath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    NSData *data;
    data = [NativeFsModule readData:filePath];
    if (!data) {
        return resolve(nil);
    }

    NSDictionary<NSString *, id> *blob =
        [NativeFsModule readDataAsBlobRef:data];
    NSString *type =
        blob[@"type"] != nil ? blob[@"type"] : @"application/octet-stream";
    NSString *text =
        [NSString stringWithFormat:@"data:%@;base64,%@", type,
                                   [data base64EncodedStringWithOptions:0]];
    resolve(text);
}

RCT_EXPORT_METHOD(fileExists
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    bool exists = [[NSFileManager defaultManager] fileExistsAtPath:filepath];
    return resolve([NSNumber numberWithBool:exists]);
}

RCT_EXPORT_METHOD(readJson
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH,
                      [NativeFsModule formatError:@"Path not accessible"],
                      error);
    }

    NSDictionary *data = [NativeFsModule readJson:filepath
                                            error:&error];
    if (error) {
        reject(ERROR_SERIALIZATION_FAILED, [NativeFsModule formatError:@"Failed to deserialize JSON"], error);
        return;
    }

    resolve(data);
}

RCT_EXPORT_METHOD(writeJson
                  : (NSDictionary<NSString *, id> *)data filepath
                  : (NSString *)filepath resolver
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject) {
    NSError *error;
    [NativeFsModule ensureWhiteListedPath:[NSArray arrayWithObject:filepath]
                                    error:&error];
    if (error) {
        return reject(INVALID_PATH, @"Path not accessible", error);
    }

    NSData *res =
        [NSJSONSerialization dataWithJSONObject:data
                                        options:NSJSONWritingPrettyPrinted
                                          error:&error];
    if (error) {
        reject(ERROR_SERIALIZATION_FAILED,
               [NativeFsModule formatError:@"Failed to serialize JSON"], error);
        return;
    }

    [NativeFsModule save:res filepath:filepath error:&error];
    if (error) {
        return reject(ERROR_SAVE_FAILED,
                      [NativeFsModule formatError:@"Failed to write JSON"],
                      error);
    }

    resolve(nil);
}

- (NSDictionary *)constantsToExport {
    return @{
        @"DocumentDirectoryPath" : [NSSearchPathForDirectoriesInDomains(
            NSDocumentDirectory, NSUserDomainMask, YES) firstObject],
        @"SUPPORTS_DIRECTORY_MOVE" : @YES,
        @"SUPPORTS_ENCRYPTION" : @YES
    };
}

@end

File: ios/Modules/NativeFsModule/NativeFsModule.swift
import Foundation
import React

@objcMembers
public class NativeFsModule: NSObject {
    
    private static var encryptionEnabled = false
    
    // Error constants
    private static let ERROR_SAVE_FAILED = "ERROR_SAVE_FAILED"
    private static let ERROR_READ_FAILED = "ERROR_READ_FAILED"
    private static let ERROR_MOVE_FAILED = "ERROR_MOVE_FAILED"
    private static let ERROR_DELETE_FAILED = "ERROR_DELETE_FAILED"
    private static let ERROR_SERIALIZATION_FAILED = "ERROR_SERIALIZATION_FAILED"
    private static let INVALID_PATH = "INVALID_PATH"
    
    private static let NativeFsErrorDomain = "com.mendix.mendixnative.nativefsmodule"
    
    public static func setEncryptionEnabled(_ enabled: Bool) {
        encryptionEnabled = enabled
    }
    
    private static func formatError(_ message: String) -> String {
        return "\(String(describing: NativeFsModule.self)): \(message)"
    }
    
    private func getBlobManager() -> RCTBlobManager? {
        guard let blobManager: RCTBlobManager = ReactAppProvider.getModule(type: RCTBlobManager.self) else {
            NSLog("NativeFsModule: Failed to get RCTBlobManager")
            return nil
        }
        return blobManager
    }
    
    private func readBlobRefAsData(_ blob: [String: Any]) -> Data? {
        guard let data = getBlobManager()?.resolve(blob) else {
            NSLog("NativeFsModule: Failed to resolve blob")
            return nil
        }
        return data
    }
    
    private func readDataAsBlobRef(_ data: Data) -> [String: Any]? {
        guard let blobId = getBlobManager()?.store(data) else {
            NSLog("NativeFsModule: Failed to store data as blob")
            return nil
        }
        return [
            "blobId": blobId as Any,
            "offset": 0,
            "length": data.count
        ]
    }
    
    static func readData(_ filePath: String) -> Data? {
        guard FileManager.default.fileExists(atPath: filePath) else {
            return nil
        }
        
        do {
            return try Data(contentsOf: URL(fileURLWithPath: filePath), options: .mappedRead)
        } catch {
            return nil
        }
    }
    
    static func readJson(_ filePath: String) throws -> [String: Any]? {
        guard let data = readData(filePath) else {
            return nil
        }
        let result = try JSONSerialization.jsonObject(with: data, options: .mutableContainers)
        return result as? [String: Any]
    }
    
    static func save(_ data: Data, filepath: String) throws {
        let directoryURL = URL(fileURLWithPath: (filepath as NSString).deletingLastPathComponent)
        try FileManager.default.createDirectory(at: directoryURL, withIntermediateDirectories: true, attributes: nil)
        let options: Data.WritingOptions = encryptionEnabled ? [.atomic, .completeFileProtection] : .atomic
        try data.write(to: URL(fileURLWithPath: filepath), options: options)
    }
    
    static func move(_ filepath: String, newPath: String) throws {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: filepath) else {
            throw NSError(domain: NativeFsErrorDomain, code: -1, userInfo: [NSLocalizedDescriptionKey: "File does not exist"])
        }
        let directoryURL = URL(fileURLWithPath: (newPath as NSString).deletingLastPathComponent)
        try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true, attributes: nil)
        try fileManager.moveItem(atPath: filepath, toPath: newPath)
    }
    
    static func remove(_ filepath: String) throws {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: filepath) else {
            NSLog("Trying to delete non-existing file: \(filepath)")
            return
        }
        try fileManager.removeItem(atPath: filepath)
    }
    
    static func ensureWhiteListedPath(_ paths: [String]) throws {
        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? ""
        let cachesPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first ?? ""
        let tempPath = (NSTemporaryDirectory() as NSString).standardizingPath
        
        for path in paths {
            if !path.hasPrefix(documentsPath) &&
                !path.hasPrefix(cachesPath) &&
                !path.hasPrefix(tempPath) {
                throw NSError(
                    domain: NativeFsErrorDomain,
                    code: 999,
                    userInfo: [NSLocalizedDescriptionKey: "The path \(path) does not point to the documents directory"]
                )
            }
        }
    }
    
    static func list(_ dirPath: String) -> [String] {
        guard let enumerator = FileManager.default.enumerator(atPath: dirPath) else {
            return []
        }
        return enumerator.allObjects as? [String] ?? []
    }
    
    // MARK: - React Native Bridge Methods
    
    public func setEncryptionEnabled(_ enabled: Bool) {
        NativeFsModule.setEncryptionEnabled(enabled)
    }
    
    public func save(
        _ blob: [String: Any],
        filepath: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        guard let data = readBlobRefAsData(blob) else {
            reject(NativeFsModule.ERROR_READ_FAILED, NativeFsModule.formatError("Failed to read blob"), nil)
            return
        }
        
        do {
            try NativeFsModule.save(data, filepath: filepath)
            resolve(nil)
        } catch {
            reject(NativeFsModule.ERROR_SAVE_FAILED, NativeFsModule.formatError("Save failed"), error)
        }
    }
    
    public func read(_ filepath: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        guard let data = NativeFsModule.readData(filepath) else {
            resolve(nil)
            return
        }
        
        guard let blob = readDataAsBlobRef(data) else {
            reject(NativeFsModule.ERROR_READ_FAILED, NativeFsModule.formatError("Failed to create blob"), nil)
            return
        }
        
        resolve(blob)
    }
    
    public func move(_ filepath: String,
                     newPath: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filepath, newPath, reject: reject) else { return }
        
        do {
            try NativeFsModule.move(filepath, newPath: newPath)
            resolve(nil)
        } catch {
            reject(NativeFsModule.ERROR_MOVE_FAILED, NativeFsModule.formatError("Failed to move file"), error)
        }
    }
    
    public func remove(_ filepath: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        do {
            try NativeFsModule.remove(filepath)
            resolve(nil)
        } catch {
            reject(NativeFsModule.ERROR_DELETE_FAILED, NativeFsModule.formatError("Failed to delete file"), error)
        }
    }
    
    public func list(_ dirPath: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(dirPath, reject: reject) else { return }
        
        resolve(NativeFsModule.list(dirPath))
    }
    
    public func readAsDataURL(_ filePath: String,
                              resolve: @escaping RCTPromiseResolveBlock,
                              reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filePath, reject: reject) else { return }
        
        guard let data = NativeFsModule.readData(filePath) else {
            resolve(nil)
            return
        }
        
        let base64String = data.base64EncodedString()
        let dataURL = "data:application/octet-stream;base64,\(base64String)"
        resolve(dataURL)
    }
    
    public func fileExists(_ filepath: String,
                           resolve: @escaping RCTPromiseResolveBlock,
                           reject: @escaping RCTPromiseRejectBlock) {
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        let exists = FileManager.default.fileExists(atPath: filepath)
        resolve(NSNumber(value: exists))
    }
    
    public func readJson(_ filepath: String,
                         resolve: @escaping RCTPromiseResolveBlock,
                         reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        do {
            let data = try NativeFsModule.readJson(filepath)
            resolve(data)
        } catch {
            reject(NativeFsModule.ERROR_SERIALIZATION_FAILED, NativeFsModule.formatError("Failed to deserialize JSON"), error)
        }
    }
    
    public func writeJson(_ data: [String: Any],
                          filepath: String,
                          resolve: @escaping RCTPromiseResolveBlock,
                          reject: @escaping RCTPromiseRejectBlock) {
        
        guard isWhiteListedPath(filepath, reject: reject) else { return }
        
        var jsonData: Data
        do {
            jsonData = try JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
        } catch {
            reject(NativeFsModule.ERROR_SERIALIZATION_FAILED, NativeFsModule.formatError("Failed to serialize JSON"), error)
            return
        }
        
        do {
            try NativeFsModule.save(jsonData, filepath: filepath)
            resolve(nil)
        } catch {
            reject(NativeFsModule.ERROR_SAVE_FAILED, NativeFsModule.formatError("Failed to write JSON"), error)
        }
    }
    
    public let constants: NSDictionary = [
        "DocumentDirectoryPath": NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? "",
        "SUPPORTS_DIRECTORY_MOVE": true,
        "SUPPORTS_ENCRYPTION": true
    ]
    
    private func isWhiteListedPath(_ paths: String..., reject: RCTPromiseRejectBlock) -> Bool {
        do {
            try NativeFsModule.ensureWhiteListedPath(paths)
            return true
        } catch let error {
            reject(NativeFsModule.INVALID_PATH, NativeFsModule.formatError("Path not accessible"), error)
            return false
        }
    }
}

